### PR TITLE
e2e test added for gcs multipart upload policy

### PR DESCRIPTION
**Pull request by neerajsinghal05**
e2e test added for gcs multipart upload policy

### CODE REVIEW

File: src/e2e-test/features/bigquery/source/BigQueryToGCS.feature [hunk 1]
```diff
@@ -116,3 +116,39 @@ Feature: BigQuery source - Verification of BigQuery to GCS successful data trans
     Then Open and capture logs
     Then Verify the pipeline status is "Succeeded"
     Then Verify data is transferred to target GCS bucket
+
+  @BQ_SOURCE_DATATYPE_TEST @GCS_CSV_MULTI_PART_UPLOAD
+Scenario:Validate successful records transfer from BigQuery to GCS
+  Given Open Datafusion Project to configure pipeline
+  When Source is BigQuery
```
Comment (itsmekumari) [inline src/e2e-test/features/bigquery/source/BigQueryToGCS.feature]: Use the latest framework pipeline steps

File: src/e2e-test/features/bigquery/source/BigQueryToGCS.feature [hunk 2]
```diff
@@ -116,3 +116,39 @@ Feature: BigQuery source - Verification of BigQuery to GCS successful data trans
...
+  When Sink is GCS
```
Comment (itsmekumari) [inline src/e2e-test/features/bigquery/source/BigQueryToGCS.feature]: Check all the steps in scenario and Use the latest framework pipeline steps

File: src/e2e-test/features/bigquery/source/BigQueryToGCS.feature [hunk 3]
```diff
@@ -116,3 +116,39 @@ Feature: BigQuery source - Verification of BigQuery to GCS successful data trans
     Then Open and capture logs
     Then Verify the pipeline status is "Succeeded"
     Then Verify data is transferred to target GCS bucket
+
+  @BQ_SOURCE_DATATYPE_TEST @GCS_CSV_MULTI_PART_UPLOAD
+Scenario:Validate successful records transfer from BigQuery to GCS
```
Comment (itsmekumari) [inline src/e2e-test/features/bigquery/source/BigQueryToGCS.feature]: Change the scenario-> Validate successful records transfer from BigQuery to GCS with bucket having delete multi part upload policy enabled

File: src/e2e-test/features/bigquery/source/BigQueryToGCS.feature [hunk 4]
```diff
@@ -116,3 +116,39 @@ Feature: BigQuery source - Verification of BigQuery to GCS successful data trans
...
+  Then Enter BigQuery property datasetProjectId "projectId"
+  Then Override Service account details if set in environment variables
+  Then Enter BigQuery property dataset "dataset"
+  Then Enter BigQuery source property table name
+  Then Validate output schema with expectedSchema "bqSourceSchemaDatatype"
+  Then Validate "BigQuery" plugin properties
+  Then Close the BigQuery properties
+  Then Open GCS sink properties
+  Then Enter the GCS sink mandatory properties
+    Then Close the GCS properties
```
Comment (itsmekumari) [inline src/e2e-test/features/bigquery/source/BigQueryToGCS.feature]: Check the indentation for the steps

File: src/e2e-test/features/bigquery/source/BigQueryToGCS.feature [hunk 5]
```diff
@@ -116,3 +116,39 @@ Feature: BigQuery source - Verification of BigQuery to GCS successful data trans
     Then Open and capture logs
     Then Verify the pipeline status is "Succeeded"
     Then Verify data is transferred to target GCS bucket
+
+  @BQ_SOURCE_DATATYPE_TEST @GCS_CSV_MULTI_PART_UPLOAD
```
Comment (itsmekumari) [inline src/e2e-test/features/bigquery/source/BigQueryToGCS.feature]: check the indentation, remove the extra space before BQ hook

File: src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java [hunk 6]
```diff
@@ -1030,4 +1033,17 @@ public static void createSinkBQExistingDatatypeTable() throws IOException, Inter
...
+    StorageClient.uploadObject(bucketName, filePath, filePath);
+    PluginPropertyUtils.addPluginProp("gcsSourceBucketName", bucketName);
+    PluginPropertyUtils.addPluginProp("gcsSourcePath", "gs://" + bucketName + "/" + filePath);
+    BeforeActions.scenario.write("Created GCS Bucket " + bucketName + " containing " + filePath + " file");
+    return bucketName;
+  }
+
+  @Before(order = 1, value = "@GCS_CSV_MULTI_PART_UPLOAD")
+  public static void createBucketWithCSVFileAND() throws IOException, URISyntaxException {
+    gcsTargetBucketName = createGCSBucketWithFileAndLifeCycle(PluginPropertyUtils.pluginProp("gcsCsvFile"));
```
Comment (itsmekumari) [inline src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java]: Why are we uploading a csv file here in bucket? the pipeline from BQ to GCS itself should create a file in bucket.

Comment (neerajsinghal05) [reply src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java]: createGCSBucketWithFileAndLifeCycle this method required the file path , so this is the CSV file path.

Comment (neerajsinghal05) [reply src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java]: in its parameter it is required.

File: src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java [hunk 7]
```diff
@@ -1030,4 +1033,17 @@ public static void createSinkBQExistingDatatypeTable() throws IOException, Inter
...
+    String bucketName = StorageClient.createBucketwithLifeCycle("00000000-e2e-" + UUID.randomUUID(),5 ).getName();
+    StorageClient.uploadObject(bucketName, filePath, filePath);
+    PluginPropertyUtils.addPluginProp("gcsSourceBucketName", bucketName);
+    PluginPropertyUtils.addPluginProp("gcsSourcePath", "gs://" + bucketName + "/" + filePath);
+    BeforeActions.scenario.write("Created GCS Bucket " + bucketName + " containing " + filePath + " file");
+    return bucketName;
+  }
+
+  @Before(order = 1, value = "@GCS_CSV_MULTI_PART_UPLOAD")
+  public static void createBucketWithCSVFileAND() throws IOException, URISyntaxException {
```
Comment (itsmekumari) [inline src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java]: Please change the method name as we are not uploading any file here.

File: src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java [hunk 8]
```diff
@@ -211,7 +214,7 @@ public static void createTargetGCSBucketWithCSVFile() throws IOException, URISyn
     BeforeActions.scenario.write("GCS target bucket name - " + gcsTargetBucketName);
   }
 
-  @After(order = 1, value = "@GCS_SINK_TEST or @GCS_SINK_EXISTING_BUCKET_TEST")
+  @After(order = 1, value = "@GCS_SINK_TEST or @GCS_SINK_EXISTING_BUCKET_TEST or @GCS_CSV_MULTI_PART_UPLOAD")
```
Comment (itsmekumari) [inline src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java]: replace CSV text with SINK text in hook name

File: src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java [hunk 9]
```diff
@@ -1030,4 +1033,17 @@ public static void createSinkBQExistingDatatypeTable() throws IOException, Inter
...
+  private static String createGCSBucketWithFileAndLifeCycle(String filePath) throws IOException, URISyntaxException {
+    String bucketName = StorageClient.createBucketwithLifeCycle("00000000-e2e-" + UUID.randomUUID(),5 ).getName();
+    StorageClient.uploadObject(bucketName, filePath, filePath);
+    PluginPropertyUtils.addPluginProp("gcsSourceBucketName", bucketName);
+    PluginPropertyUtils.addPluginProp("gcsSourcePath", "gs://" + bucketName + "/" + filePath);
+    BeforeActions.scenario.write("Created GCS Bucket " + bucketName + " containing " + filePath + " file");
+    return bucketName;
+  }
+
+  @Before(order = 1, value = "@GCS_CSV_MULTI_PART_UPLOAD")
```
Comment (itsmekumari) [inline src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java]: replace CSV text with SINK text in hook name

**Review by rahuldash171 — APPROVED** [review]

File: src/e2e-test/features/bigquery/source/BigQueryToGCS.feature [hunk 10]
```diff
@@ -116,3 +116,41 @@ Feature: BigQuery source - Verification of BigQuery to GCS successful data trans
...
     Then Verify data is transferred to target GCS bucket
+
+  @BQ_SOURCE_DATATYPE_TEST @GCS_SINK_MULTI_PART_UPLOAD
+Scenario:Validate successful records transfer from BigQuery to GCS with bucket having delete multi part upload policy enabled
+  Given Open Datafusion Project to configure pipeline
+  When Expand Plugin group in the LHS plugins list: "Source"
+  When Select plugin: "BigQuery" from the plugins list as: "Source"
+  When Expand Plugin group in the LHS plugins list: "Sink"
+  When Select plugin: "GCS" from the plugins list as: "Sink"
+  Then Open BigQuery source properties
```
Comment (itsmekumari) [inline src/e2e-test/features/bigquery/source/BigQueryToGCS.feature]: Use new framework steps for all these like navigating into plugin properties, entering reference name

File: src/e2e-test/features/bigquery/source/BigQueryToGCS.feature [hunk 11]
```diff
@@ -116,3 +116,41 @@ Feature: BigQuery source - Verification of BigQuery to GCS successful data trans
...
+  Then Enter BigQuery property reference name
+  Then Enter BigQuery property projectId "projectId"
+  Then Enter BigQuery property datasetProjectId "projectId"
+  Then Override Service account details if set in environment variables
+  Then Enter BigQuery property dataset "dataset"
+  Then Enter BigQuery source property table name
```
Comment (itsmekumari) [inline src/e2e-test/features/bigquery/source/BigQueryToGCS.feature]: Use new framework steps, for entering all the field properties. Check the new tests added for BQ or GCS.

File: src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java [hunk 12]
```diff
@@ -1030,4 +1030,14 @@ public static void createSinkBQExistingDatatypeTable() throws IOException, Inter
     PluginPropertyUtils.addPluginProp(" bqTargetTable",  bqTargetTable);
     BeforeActions.scenario.write("BQ Target Table " +  bqTargetTable + " updated successfully");
   }
+  private static String createGCSBucketLifeCycle() throws IOException, URISyntaxException {
+    String bucketName = StorageClient.createBucketwithLifeCycle("00000000-e2e-" + UUID.randomUUID(), 30).getName();
+    PluginPropertyUtils.addPluginProp("gcsSourceBucketName", bucketName);
```
Comment (itsmekumari) [inline src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java]: Since it is Sink, This should be a 'gcsTargetBucketName'

Comment (itsmekumari) [inline src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java]: Also, where is gcs target path picked from?

File: src/e2e-test/features/bigquery/source/BigQueryToGCS.feature (hunk 10 above)
Comment (neerajsinghal05) [reply src/e2e-test/features/bigquery/source/BigQueryToGCS.feature]: Done , changes added as per new framework.

File: src/e2e-test/features/bigquery/source/BigQueryToGCS.feature (hunk 11 above)
Comment (neerajsinghal05) [reply src/e2e-test/features/bigquery/source/BigQueryToGCS.feature]: Done , changes added as per new framework.

File: src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java (hunk 12 above)
Comment (neerajsinghal05) [reply src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java]: Done , Changed it to gcsTargetBucketName.

Comment (neerajsinghal05) [reply src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java]: enterGCSSinkPropertyPath from this step the path is getting.

File: src/e2e-test/features/bigquery/source/BigQueryToGCS.feature [hunk 13]
```diff
@@ -124,33 +124,33 @@ Scenario:Validate successful records transfer from BigQuery to GCS with bucket h
...
-  Then Validate output schema with expectedSchema "bqSourceSchemaDatatype"
+  And Enter input plugin property: "table" with value: "bqSourceTable"
+  Then Click on the Get Schema button
   Then Validate "BigQuery" plugin properties
   Then Close the BigQuery properties
-  Then Open GCS sink properties
-  Then Enter the GCS sink mandatory properties
-  Then Close the GCS properties
+  Then Navigate to the properties page of plugin: "GCS"
+  Then Enter input plugin property: "referenceName" with value: "sourceRef"
```
Comment (itsmekumari) [inline src/e2e-test/features/bigquery/source/BigQueryToGCS.feature:139]: Add the step to enter project id

Comment (neerajsinghal05) [reply src/e2e-test/features/bigquery/source/BigQueryToGCS.feature:139]: Changes Added as per the review comment.

File: src/e2e-test/features/bigquery/source/BigQueryToGCS.feature (hunk 1 above)
Comment (neerajsinghal05) [reply src/e2e-test/features/bigquery/source/BigQueryToGCS.feature]: Changes Added as per the Suggestion.

File: src/e2e-test/features/bigquery/source/BigQueryToGCS.feature (hunk 2 above)
Comment (neerajsinghal05) [reply src/e2e-test/features/bigquery/source/BigQueryToGCS.feature]: Changes Added as per the Suggestion.

File: src/e2e-test/features/bigquery/source/BigQueryToGCS.feature (hunk 3 above)
Comment (neerajsinghal05) [reply src/e2e-test/features/bigquery/source/BigQueryToGCS.feature]: Changes Added as per the Suggestion.

File: src/e2e-test/features/bigquery/source/BigQueryToGCS.feature (hunk 4 above)
Comment (neerajsinghal05) [reply src/e2e-test/features/bigquery/source/BigQueryToGCS.feature]: Changes Added as per the Suggestion.

File: src/e2e-test/features/bigquery/source/BigQueryToGCS.feature (hunk 5 above)
Comment (neerajsinghal05) [reply src/e2e-test/features/bigquery/source/BigQueryToGCS.feature]: Changes Added as per the Suggestion.

File: src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java (hunk 6 above)
Comment (neerajsinghal05) [reply src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java]: Changes Added as per the Suggestion.

File: src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java (hunk 7 above)
Comment (neerajsinghal05) [reply src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java]: Changes Added as per the Suggestion.

File: src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java (hunk 8 above)
Comment (neerajsinghal05) [reply src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java]: Changes Added as per the Suggestion.

File: src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java (hunk 9 above)
Comment (neerajsinghal05) [reply src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java]: Changes Added as per the Suggestion.

**Review by itsmekumari — APPROVED** [review]

**Pull request merged**